cmd/exo/cmd/firewall_add.go

```diff
@@ -32,103 +37,275 @@ func firewallAddRun(cmd *cobra.Command, args []string) {
+		portsRange[i].end = uint16(p1)
+		if len(pRange) == 2 {
+			p2, err := strconv.ParseUint(pRange[0], 10, 16)
+			if err != nil {
+				return nil, err
+			}
+			portsRange[i].end = uint16(p2)
+		}
+	}
+	return portsRange, nil
 }
 
-func getMyCIDR(isIpv6 bool) string {
+func getSGs(cs *egoscale.Client, sgs []string) ([]egoscale.UserSecurityGroup, error) {
+
+	sgRes := []egoscale.UserSecurityGroup{}
 
-	cidrSuffix := ""
-	dnsServer := ""
+	for _, sg := range sgs {
 
+		s, err := getSecuGrpWithNameOrID(cs, sg)
+		if err != nil {
+			return nil, err
+		}
+		sgRes = append(sgRes, egoscale.UserSecurityGroup{Group: s.Name, Account: s.Account})
+	}
+	return sgRes, nil
+}
+
+func getDefaultRule(ruleName string, isIpv6 bool) (*egoscale.AuthorizeSecurityGroupIngress, error) {
+
+	cidr := defaultCidr
 	if isIpv6 {
-		dnsServer = "resolver1.ipv6-sandbox.opendns.com"
-		cidrSuffix = "/128"
-	} else {
-		dnsServer = "resolver1.opendns.com"
-		cidrSuffix = "/32"
+		cidr = defaultCidr6
 	}
-	resolver := net.Resolver{
-		Dial: func(ctx context.Context, network, address string) (net.Conn, error) {
-			return net.Dial("udp", dnsServer+":53")
-		},
-		PreferGo: true,
+
+	ruleName = strings.ToLower(ruleName)
+	if ruleName == "ping" {
+		return &egoscale.AuthorizeSecurityGroupIngress{
+			Protocol:    "icmp",
+			CidrList:    []string{cidr},
+			IcmpType:    8,
+			IcmpCode:    0,
+			Description: "",
+		}, nil
 	}
 
-	ip, err := resolver.LookupIPAddr(context.Background(), "myip.opendns.com")
-	if err != nil {
-		log.Fatal(err)
+	for d := daytime; d <= minecraft; d++ {
+		if port.String(d) == ruleName {
+			return &egoscale.AuthorizeSecurityGroupIngress{
+				Protocol:    "tcp",
+				CidrList:    []string{cidr},
+				StartPort:   uint16(d),
+				EndPort:     uint16(d),
+				Description: fmt.Sprintf(""),
+			}, nil
+		}
 	}
 
-	if len(ip) < 1 {
-		return ""
+	return nil, fmt.Errorf("default rule %q not found", ruleName)
+}
+
+func addRule(rule *egoscale.AuthorizeSecurityGroupIngress, isEgress bool) error {
+	var err error
+	if isEgress {
+		_, err = cs.Request((*egoscale.AuthorizeSecurityGroupEgress)(rule))
+	} else {
+		_, err = cs.Request(rule)
 	}
 
-	return (ip[0].IP.String() + cidrSuffix)
+	if err != nil {
+		return err
+	}
 
+	return nil
 }
 
 func init() {
 	firewallAddCmd.Run = firewallAddRun
 
-	firewallAddCmd.Flags().BoolP("ipv6", "6", false, "Add rule for any IPv6 source")
-	firewallAddCmd.Flags().BoolP("my-ip", "m", false, "Add rule only for my IP as a source")
+	firewallAddCmd.Flags().BoolP("ipv6", "6", false, "Set ipv6 on default rules or on --my-ip")
+	firewallAddCmd.Flags().BoolP("my-ip", "", false, "Set CIDR for my ip")
+	firewallAddCmd.Flags().BoolP("engress", "e", false, "By default rule is INGRESS (set --engress to have ENGRESS rule)")
+	firewallAddCmd.Flags().StringP("protocol", "p", "", "Rule Protocol available [tcp, udp, icmp, icmpv6, ah, esp, gre]")
+	firewallAddCmd.Flags().StringP("cidr", "c", "", "Rule Cidr [CIDR 0.0.0.0/0,192.168.0.0/16...]")
```

Comment:
`::/0` is probably more useful than 192.168.x.x (a non-routable IP)

Reply:
yes